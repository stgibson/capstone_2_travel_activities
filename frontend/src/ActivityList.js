import React from "react";
import { useSelector, useDispatch, shallowEqual } from "react-redux";
import Activity from "./Activity";
import {
  addFavoriteActivityToAPI,
  removeFavoriteActivityFromAPI
} from "./actions/activities";

/**
 * Component for displaying activity results from searching activities
 * @param {Object} param0 
 * @returns JSX code for rendering list of activities
 */
const ActivityList = ({ title, activities }) => {
  const token = useSelector(store => store.token);
  const likedActivities = useSelector(store => store.activities, shallowEqual);
  const dispatch = useDispatch();

  /**
   * Calls dispatch to add activity to list of currently logged in user's
   * favorite activities
   * @param {Object} evt 
   */
  const handleLike = evt => {
    const { id } = evt.target.parentElement;
    dispatch(addFavoriteActivityToAPI(Number.parseInt(id), token));
  };

  /**
   * Calls dispatch to remove activity from list of currently logged in user's
   * favorite activities
   * @param {Object} evt 
   */
  const handleUnlike = evt => {
    const { id } = evt.target.parentElement;
    dispatch(removeFavoriteActivityFromAPI(Number.parseInt(id), token));
  }

  /**
   * Determines if currently logged in user likes activity with provided id
   * @param {number} id 
   * @returns true if user likes activity, false otherwise
   */
  const liked = id => {
    if (likedActivities[id]) {
      return true;
    }
    return false;
  };

  return (
    <div>
      <h2>{ title }</h2>
      {
        activities.map(activity => (
          liked(activity.id) ? (
            <Activity
              key={ activity.id }
              activity={ activity }
              btnText="Unlike"
              btnCallback={ evt => handleUnlike(evt) }
            />
          ) : (
            <Activity
              key={ activity.id }
              activity={ activity }
              btnText="Like"
              btnCallback={ evt => handleLike(evt) }
            />
          )
        ))
      }
    </div>
  );
};

export default ActivityList;