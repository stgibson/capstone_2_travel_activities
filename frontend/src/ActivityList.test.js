import React from "react";
import { MemoryRouter } from "react-router-dom";
import { createStore } from "redux";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import ActivityList from "./ActivityList";
import rootReducer from "./reducers/root";

const activities = [
  { id: 1, name: "Test Activity 1" },
  { id: 2, name: "Test Activity 2" }
];

let store;
beforeEach(() => {
  store = createStore(rootReducer);
});

it("renders without crashing", () => {
  render(
    <Provider store={ store }>
      <MemoryRouter>
        <ActivityList title="Activity List" activities={ activities } />
      </MemoryRouter>
    </Provider>
  );
});

it("matches snapshot", () => {
  const { asFragment } = render(
    <Provider store={ store }>
      <MemoryRouter>
        <ActivityList title="Activity List" activities={ activities } />
      </MemoryRouter>
    </Provider>
  );
  
  expect(asFragment()).toMatchSnapshot();
});