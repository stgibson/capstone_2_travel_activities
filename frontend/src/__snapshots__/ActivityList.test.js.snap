// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Activity List
    </h2>
    <div
      class="Activity"
      id="1"
    >
      <a
        class="Activity-link"
        href="/activities/1"
      >
        Test Activity 1
      </a>
      <button
        class="btn btn-primary btn-sm"
        type="button"
      >
        Like
      </button>
    </div>
    <div
      class="Activity"
      id="2"
    >
      <a
        class="Activity-link"
        href="/activities/2"
      >
        Test Activity 2
      </a>
      <button
        class="btn btn-primary btn-sm"
        type="button"
      >
        Like
      </button>
    </div>
  </div>
</DocumentFragment>
`;
